### PR TITLE
Reject negative patience at cfg

### PR DESCRIPTION
**Pull request by wizzseen**
## 🛠️ Issue
Follow-up to #26208-style cfg validation. Several overrides are accepted today but cause silent wrong training or deep crashes:
- `batch=0` / `batch<=-2` → unintended AutoBatch
- `patience<0` → early-stop every epoch
- `time<=0` → ignored or instant stop
- non-finite floats (e.g. `warmup_epochs=nan`) → OverflowError later
- `workers<0` → DataLoader error deep in setup

## 🔄 Change
Reject these in `check_cfg`. Keep legal edges: `batch=-1`, `0<batch<1`, `patience=0`, `workers=0`.

## 📋 Evidence
Unit test: `tests/test_python.py::test_cfg_rejects_fuzzed_values`

## Checklist
- [x] Minimal change (cfg + default.yaml comment + tests)
- [x] CLA will be signed if prompted

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Updated configuration validation to reject invalid batch, patience, time, workers, and non-finite float values before training reaches deeper setup failures.

### 📊 Key Changes
- Extended `check_cfg` validation for finite float values, valid batch modes, and positive `time` values.
- Added regression coverage for invalid and non-finite overrides, including zero or negative batch, patience, time, and workers values.
- Confirmed supported edge cases such as AutoBatch with `batch=-1`, fractional batch sizes, `patience=0`, and `workers=0`.
- Clarified the default batch configuration comment and documented the patience and workers defaults.

### 🎯 Purpose & Impact
Invalid configuration overrides now fail early with `TypeError` or `ValueError`, preventing unintended AutoBatch behavior, immediate early stopping, invalid time limits, non-finite-value crashes, and deep DataLoader failures.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @wizzseen, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Reviewed all 3 changed files and the surrounding cfg, trainer, dataloader, and test paths. The new validation correctly covers non-finite scalar floats, non-positive time, negative patience, negative workers, and the documented batch edge cases, but it still allows fractional batch sizes at or above 1, which fail later in training/export. One fix is needed.

<details><summary>💬 Posted 1 inline comment</summary>

- 💡 **MEDIUM** `ultralytics/cfg/__init__.py:442` This condition still accepts non-integral batch sizes >= 1, such as `batch=1.5`, despite the documented contract requiring an integer. Training passes this value into `build_dataloader` and exporter paths use it as a tensor dimension, causing a deeper `TypeError` instead of a cfg validation error. Require the `>= 1` branch to be integral.
</details>

File: ultralytics/cfg/__init__.py [hunk 1]
```diff
@@ -426,13 +429,23 @@ def check_cfg(cfg: dict, hard: bool = True) -> None:
...
+                if not isinstance(v, FLOAT_OR_INT):
+                    if hard:
+                        raise TypeError(
+                            f"'{k}={v}' is of invalid type {type(v).__name__}. "
+                            f"Valid '{k}' types are int (i.e. '{k}=0') or float (i.e. '{k}=0.5')"
+                        )
+                    cfg[k] = v = float(v)
+                if not math.isfinite(v):
+                    raise ValueError(f"'{k}={v}' is an invalid value. '{k}' must be finite.")
+                if k == "batch" and not (v == -1 or 0 < v < 1 or v >= 1):
```
Comment (UltralyticsAssistant) [inline ultralytics/cfg/__init__.py]: 💡 **MEDIUM**: This condition still accepts non-integral batch sizes >= 1, such as `batch=1.5`, despite the documented contract requiring an integer. Training passes this value into `build_dataloader` and exporter paths use it as a tensor dimension, causing a deeper `TypeError` instead of a cfg validation error. Require the `>= 1` branch to be integral.

**Suggested change:**
```suggestion
                if k == "batch" and not (v == -1 or 0 < v < 1 or (v >= 1 and v == int(v))):
```

Comment (wizzseen) [reply ultralytics/cfg/__init__.py]: Applied — batch >= 1 now requires an integral value, added coverage for batch=1.5.

**Comment by wizzseen** [comment]
I have read the CLA Document and I sign the CLA

**Comment by glenn-jocher** [comment]
Thanks @wizzseen. I ran each of these values on `main` through real training before deciding what to keep:

- `patience=-1`: trains one epoch and stops with `no improvement observed in last -1 epochs`. This is the one silent failure, and `-1` reads like "disabled" next to `batch=-1` and `save_period=-1`. Kept, as a single entry in the existing `CFG_INT_MIN` table.
- `time=0`: trains normally, because a falsy `time` means no time limit just like the `None` default, so rejecting it would break working configs.
- `batch=0` and `batch=-2`: take the logged AutoBatch path and train fine.
- `workers=-2`: ignored on CPU and MPS, and on CUDA torch already reports `num_workers option should be non-negative`.
- `warmup_epochs=nan`: trains with finite losses, and the reported `OverflowError` did not reproduce.

I pushed a reduction to the `patience` table entry, one line in the existing fuzz test, and the clearer `batch` comment in `default.yaml`. The bespoke `batch`, `time` and `isfinite` branches in `check_cfg` are removed, since we avoid adding validation for inputs that already fail clearly or work.

**Review by glenn-jocher — APPROVED** [review]
Reduced to the `patience` minimum in the existing `CFG_INT_MIN` table after testing each value on `main`. LGTM.

**Comment by wizzseen** [comment]
> Reduced to the `patience` minimum in the existing `CFG_INT_MIN` table after testing each value on `main`. LGTM.

Thanks for testing these on main and narrowing the scope — looks good to me.
